spmd/tensor/redistribute.py

```diff
@@ -100,53 +100,57 @@ def redistribute_spmd_tensor(
+                # TODO: support uneven sharding
+                output_tensor = torch.empty(
+                    new_size,
+                    dtype=local_tensor.dtype,
+                    layout=local_tensor.layout,
+                    device=local_tensor.device,
+                )
+
+                new_local_tensor = device_mesh.all_gather(
+                    output_tensor,
                     local_tensor,
                     mesh_dim=i,
+                    tensor_dim=shard_spec.dim,
                 )
-                new_local_tensor = torch.cat(gathered_list, dim=shard_spec.dim)
         elif target.is_shard():
             # Case 2: target is Shard
-            shard_dim = target.dim  # type: ignore
+            target_dim = target.dim  # type: ignore
             num_chunks = device_mesh.size(dim=i)
-            assert (
-                input.size(shard_dim) % num_chunks == 0
-            ), f"Only support chunk sharding evenly now. (When sharding tensor of shape {input.shape} on dim={shard_dim} into {num_chunks} shards.)"
-
             chunk_size = (
-                local_tensor.size(shard_dim) // num_chunks
+                local_tensor.size(target_dim) // num_chunks
             )  # this may already be sharded on a differernt mesh dimension
             my_rank = device_mesh.get_coordinate_on_dim(dim=i)
             assert (
                 my_rank is not None
-            ), "Rank is not part of the mesh"  # TODO: figure out behavior here
+            ), f"Rank: {device_mesh.get_rank()} is not part of the mesh!"  # TODO: figure out behavior here
             if current.is_partial():
                 # reduce scatter the current tensors
-                new_tensor_size = list(local_tensor.size())
-                new_tensor_size[shard_dim] = chunk_size
-                new_local_tensor = torch.empty(
-                    new_tensor_size,
-                    device=local_tensor.device,
-                    dtype=input.dtype,
+                new_local_tensor = device_mesh.reduce_scatter(
+                    local_tensor, mesh_dim=i, tensor_dim=target_dim
                 )
-                new_local_tensor = device_mesh.reduce_scatter_base(
-                    new_local_tensor, local_tensor, mesh_dim=i
-                )
-            else:
-                assert current.is_replicate()
+            elif current.is_replicate():
                 # slice/narrow the tensor to corresponding local shard then return shard tensor
-                # TODO: support multi dim redistribute require to not use get_rank, as it would
-                # get global_rank and in the case of multidim mesh it's not correct!
                 new_local_tensor = local_tensor.narrow(
-                    shard_dim, my_rank * chunk_size, chunk_size
+                    target_dim, my_rank * chunk_size, chunk_size
                 )
+            else:
+                assert current.is_shard()
```

Comment:
Are we removing support for this? This will probably make view_ops tests fail. Do we need to remove it?

Reply:
(As discussed offline , this case shouldn't hit since _decompose_sharding above should decompose Shard(0) -> Shard(1) into Shard(0) -> Replicate -> Shard(1). We can make this a logic error, and add a comment that it shouildn't happen.  The TODO to make it all_to_all still applies.

Reply:
updated with comments.